### ISSUE TITLE
Pyright does not handle NoReturn in async await 

### ISSUE THREAD

**Issue by kumaraditya303** [microsoft/pyright#2598]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
Pyright does not handle NoReturn in async await 

**Expected behavior**
No error.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
# pyright: strict
import asyncio
from typing import NoReturn


async def func() -> NoReturn:
    while True:
        await asyncio.sleep(1)
        print("Hello World!")


async def main() -> NoReturn:
    await func()
    # <--- This will never execute 


asyncio.run(main())

```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

```console
pyright 1.1.189
```

**Comment by kumaraditya303** [comment]
Same happens with generators too:

```py
# pyright: strict
from typing import Generator, NoReturn


def func() -> Generator[int, None, NoReturn]:
    c = 0
    while True:
        c += 1
        yield c


def main() -> NoReturn:
    for i in func():
        print(i)
    # <--- This will never execute 

main()

```

**Comment by erictraut** [comment]
I've fixed the issue with async functions.

I'm not convinced the second one is valid though. The `NoReturn` type is not supposed to be used as a type argument. Mypy generates the same error in this case. I think you'll need to use a workaround in the case of a generator.

**Comment by kumaraditya303** [comment]
> I'm not convinced the second one is valid though. The `NoReturn` type is not supposed to be used as a type argument. Mypy generates the same error in this case. I think you'll need to use a workaround in the case of a generator.

If that's the case then why in async function the type can be `() -> Coroutine[Any, Any, NoReturn]` but in generator it cannot be `() -> Generator[int, None, NoReturn]`  ?

**Comment by erictraut** [comment]
`NoReturn` should not be used as a type argument, so `Coroutine[Any, Any, NoReturn]` is technically not permitted, but it is legal to use `async func() -> NoReturn`.

**Comment by kumaraditya303** [comment]
It still looks wrong to me that `NoReturn` cannot be used in case of generators, I have started a discussion on it in typing repo 
https://github.com/python/typing/discussions/960

**Comment by erictraut** [comment]
This is included in pyright 1.1.190, which I just published. It will also be included in the next release of pylance.

**Comment by BRIZINGR007** [comment]
Hi Guys , You can creaate a asyncio.create_task and await that task. With this the Pyright does not throw  <"NoReturn" is not awaitable>.

response = asyncio.create_task( asynchronous function )
response = await response
